### DISCUSSION TITLE
Variable declaration order & constraint order affecting convergence - why?

### DISCUSSION THREAD

**Discussion by edwinreuvers** [casadi/casadi#3705] [Q&A]
Hi all, 

I am playing around with CasADi, but I have stumbled upon an issue I do not understand. It seems that variable declaration order and constraint order affect the convergence of the optimization problem. 

For example, I transformed the Rocket Launch GEKKO example (see: https://apmonitor.com/wiki/index.php/Apps/RocketLaunch) into CasADi. I reformulated the dynamics a bit to have a linear set of equations (Ax=b, with x being the state-derivatives). This all works fine using this code:

```
import casadi as ca
import numpy as np
import matplotlib.pyplot as plt

def modelEq(x, u):
    v = x[1]
    mass = x[2]

    # A & b of the system xdot = A*x
    A = ca.MX.zeros(3, 3)
    A[0, 0] = 1
    A[1, 1] = mass
    A[2, 2] = 1

    b = ca.vertcat(
        v,
        u - 0.2 * v**2,
        -0.01 * u**2
    ) 
    return A, b

# Declare MX (CasADi)
x = ca.MX.sym('x', 3)
u = ca.MX.sym('u', 1)

# Model dynamics
A, b = modelEq(x, u)
xdot = ca.solve(A, b)

f_dyn = ca.Function('f_dyn', [x, u], [xdot])

# Parameters
N = 100  # number of control intervals
T = 1.0  # normalised final-time
dt = T / N

# CasADi setup
opti = ca.Opti()

# Decision variables
u_k = opti.variable(N)  # control vector
tf = opti.variable()  # final time

# State variables
s_k = opti.variable(1,N+1)
v_k = opti.variable(1,N+1)
m_k = opti.variable(1,N+1)
x_k = ca.vertcat(s_k,v_k,m_k)

# Initial guesses
opti.set_initial(s_k, np.linspace(0, 10, N+1))
opti.set_initial(v_k, np.zeros(N+1))
opti.set_initial(m_k, np.ones(N+1))
opti.set_initial(u_k, 0)
opti.set_initial(tf, 1.0)

# Boundary constraints
opti.subject_to(s_k[0] == 0)
opti.subject_to(v_k[0] == 0)
opti.subject_to(m_k[0] == 1)
opti.subject_to(s_k[N] == 10.0)
opti.subject_to(v_k[N] == 0.0)

# Box constraints
opti.subject_to(0 <= v_k)
opti.subject_to(v_k <= 1.7)
opti.subject_to(0.2 <= m_k)
opti.subject_to(-1.1 <= u_k)
opti.subject_to(u_k <= 1.1)

# Dynamic constraints
for k in range(N): 
    # Collect states at k and k+1
    Xk = x_k[:, k]
    Xk_next = x_k[:, k+1]
    
    # Collect controls at k
    Uk = u_k[k]
    
    opti.subject_to(Xk_next == Xk + dt * tf * f_dyn(Xk,Uk))

# Objective: minimize final time
opti.minimize(tf)

# Solver settings
opti.solver('ipopt',)

# Solve
sol = opti.solve()

# Extract solution
x_sol = sol.value(x_k).T
u_sol = sol.value(u_k).T
tf_sol = sol.value(tf)

time = np.linspace(0, 1, N+1)*tf_sol

# Plot results
plt.figure(1)
plt.subplot(4, 1, 1)
plt.plot(time, x_sol[:,0], 'r-', lw=2)
plt.ylabel('Position')

plt.subplot(4, 1, 2)
plt.plot(time, x_sol[:,1], 'b-', lw=2)
plt.ylabel('Velocity')

plt.subplot(4, 1, 3)
plt.plot(time, x_sol[:,2], 'k-', lw=2)
plt.ylabel('Mass')

plt.subplot(4, 1, 4)
plt.plot(time[:-1], u_sol, 'g-', lw=2)
plt.ylabel('Force')

plt.xlabel('Time')
plt.show()
```

Then I was updating my code and decided to swap the declaration of the state variables (s_k, v_k, m_k) and decision variables (u_k and tf), like this:

```
# State variables
s_k = opti.variable(1,N+1)
v_k = opti.variable(1,N+1)
m_k = opti.variable(1,N+1)
x_k = ca.vertcat(s_k,v_k,m_k)

# Decision variables
u_k = opti.variable(N)  # control vector
tf = opti.variable()  # final time
```

When I used this (i.e., the declaration of state variables before the decision variables), the solver fails to converge. Also, shuffling around with the order of constraints (i.e., first the dynamic constraint then boundary&box constraints) influence the convergence. 

In sum, my observation is that both the variable declaration order and the constraint order affect the convergence. I would like to understand why this happens. 

Thanks in advance. :rocket:

**Comment by sandeep026** [comment]
It could be due to ill-conditioning. The following works

1. change linear solver to spral for IPOPT with your guess
2.  use the following guess with default linear solver mumps
```
opti.set_initial(s_k, np.linspace(0, 5, N+1))
opti.set_initial(v_k, np.linspace(0, 1.7, N+1))
opti.set_initial(m_k, np.linspace(0.2, 10, N+1))
```
Both of these worked and hope that helps.

**Reply by edwinreuvers** [reply]
Thanks for your reply, but the problem is not that it does not converge, but my problem is that it does converge depending on the variable declaration/constraint order. This is odd because it has nothing to do with the solver and/or ill-conditioning; in both situations I use exactly the solver and initial guesses. I observe the following:

- At iteration number 0 the values for variables, objective and constraint violations are the same. 
- However, at iteration 1, the values for variables, objective and constraint violations are slightly (!) different. 
- From iteration 2 onwards, I understand that the values for variables, objective and constraint violations are different because the constraint violations, derivatives, etc. are different from iteration 1 onwards. 

Now the question is, why are the values for the variables different at iteration 1, when they are similar at iteration 0 (and therefore all the derivatives should be similar)?

**Reply by jaeandersson** [reply]
The order in which you declare variables and constraints matters. In the end, IPOPT will receive an NLP of the form 
```
minimize f(x)
subject to lbx <= x<= ubx,  lbg <= g(x)<= ubg
```
With a different order of declarations, you reshuffle the components in x and g.

In particular, this means that the Jacobian dg/dx and the Hessian d^2(f+dot(g,lam_g))/dx^2 will get reshuffled rows and columns. CasADi's AD will be slightly faster with a "good" sparsity pattern for Jacobians and Hessians, but it usually shouldn't matter much. It can matter a lot for the sparse linear solver however. They will have a "preordering" step to get a good ordering of rows and columns, sometimes implemented using METIS iirc.

It's usually a good idea to order your variables and constraint declarations so that your patterns get a block banded structure. If you have an optimal control problem, it means creating a for loop for time horizon and then declaring variables and constraints inside this for loop. Not sure if you can do that if you use the Opti abstraction (@jgillis would know), but if not, it's important that the linear solver you use has good preordering.

That being said, changing the declaration order can also expose a bug in CasADi. A quick check would be to make sure ipopt reports the same number of nonzeros in Jacobians and Hessians.

**Comment by sandeep026** [comment]
To quantify my response earlier, @edhmr  I modified your code. The code plots condition number of the NLP jacobian (eq. 2 of [ref)](https://elib.dlr.de/93327/1/Performance_analysis_of_linear_and_nonlinear_techniques.pdf).

Consider cases where guess for velocity is all zeros and all ones. This is what condition number at each iteration looks like and I have set max iteration to 100. 

![image](https://github.com/casadi/casadi/assets/16457676/ca6ee3a9-1aba-45b7-910e-39b7ef64db76)

![image](https://github.com/casadi/casadi/assets/16457676/4d3a0bc3-d4fe-4783-a948-e144d21827d0)

```
import casadi as ca
import numpy as np
import matplotlib.pyplot as plt

def modelEq(x, u):
    v = x[1]
    mass = x[2]

    # A & b of the system xdot = A*x
    A = ca.MX.zeros(3, 3)
    A[0, 0] = 1
    A[1, 1] = mass
    A[2, 2] = 1

    b = ca.vertcat(
        v,
        u - 0.2 * v**2,
        -0.01 * u**2
    ) 
    return A, b

######
vg=0


# Declare MX (CasADi)
x = ca.MX.sym('x', 3)
u = ca.MX.sym('u', 1)

# Model dynamics
A, b = modelEq(x, u)
xdot = ca.solve(A, b)

f_dyn = ca.Function('f_dyn', [x, u], [xdot])

# Parameters
N = 10  # number of control intervals
T = 1.0  # normalised final-time
dt = T / N

# CasADi setup
opti = ca.Opti()

# Decision variables
u_k = opti.variable(N)  # control vector
tf = opti.variable()  # final time

# State variables
s_k = opti.variable(1,N+1)
v_k = opti.variable(1,N+1)
m_k = opti.variable(1,N+1)
x_k = ca.vertcat(s_k,v_k,m_k)

# Initial guesses
opti.set_initial(s_k, np.linspace(0, 10, N+1))
opti.set_initial(v_k, vg+np.zeros(N+1))
opti.set_initial(m_k, np.ones(N+1))
opti.set_initial(u_k, 0)
opti.set_initial(tf, 1.0)

# Boundary constraints
opti.subject_to(s_k[0] == 0)
opti.subject_to(v_k[0] == 0)
opti.subject_to(m_k[0] == 1)
opti.subject_to(s_k[N] == 10.0)
opti.subject_to(v_k[N] == 0.0)

# Box constraints
opti.subject_to(0 <= v_k)
opti.subject_to(v_k <= 1.7)
opti.subject_to(0.2 <= m_k)
opti.subject_to(-1.1 <= u_k)
opti.subject_to(u_k <= 1.1)

# Dynamic constraints
for k in range(N): 
    # Collect states at k and k+1
    Xk = x_k[:, k]
    Xk_next = x_k[:, k+1]
    
    # Collect controls at k
    Uk = u_k[k]
    
    opti.subject_to(Xk_next == Xk + dt * tf * f_dyn(Xk,Uk))

# Objective: minimize final time
opti.minimize(tf)

# Solver settings
opti.solver('ipopt',{'ipopt.max_iter':100})

(h, g) = ca.hessian(opti.f, opti.x)
j = ca.jacobian(opti.g, opti.x)
plt.subplot(1, 2, 1)
opti.callback(lambda i: plt.semilogy(i,ca.np.linalg.cond(opti.debug.value(ca.vertcat(g.T, j)).todense()),'o'))


# Solve
sol = opti.solve_limited()
plt.title('U,X- V_guess='+str(vg))
#plt.show()
#plt.figure()
plt.grid(True)


##################################################################################
#variable reorder
##################################################################################
del opti

# Declare MX (CasADi)
x = ca.MX.sym('x', 3)
u = ca.MX.sym('u', 1)

# Model dynamics
A, b = modelEq(x, u)
xdot = ca.solve(A, b)

f_dyn = ca.Function('f_dyn', [x, u], [xdot])

# Parameters
N = 100  # number of control intervals
T = 1.0  # normalised final-time
dt = T / N

# CasADi setup
opti = ca.Opti()

# State variables
s_k = opti.variable(1,N+1)
v_k = opti.variable(1,N+1)
m_k = opti.variable(1,N+1)
x_k = ca.vertcat(s_k,v_k,m_k)

# Decision variables
u_k = opti.variable(N)  # control vector
tf = opti.variable()  # final time

# Initial guesses
opti.set_initial(s_k, np.linspace(0, 10, N+1))
opti.set_initial(v_k, vg+np.zeros(N+1))
opti.set_initial(m_k, np.ones(N+1))
opti.set_initial(u_k, 0)
opti.set_initial(tf, 1.0)

# Boundary constraints
opti.subject_to(s_k[0] == 0)
opti.subject_to(v_k[0] == 0)
opti.subject_to(m_k[0] == 1)
opti.subject_to(s_k[N] == 10.0)
opti.subject_to(v_k[N] == 0.0)

# Box constraints
opti.subject_to(0 <= v_k)
opti.subject_to(v_k <= 1.7)
opti.subject_to(0.2 <= m_k)
opti.subject_to(-1.1 <= u_k)
opti.subject_to(u_k <= 1.1)

# Dynamic constraints
for k in range(N): 
    # Collect states at k and k+1
    Xk = x_k[:, k]
    Xk_next = x_k[:, k+1]
    
    # Collect controls at k
    Uk = u_k[k]
    
    opti.subject_to(Xk_next == Xk + dt * tf * f_dyn(Xk,Uk))

# Objective: minimize final time
opti.minimize(tf)

# Solver settings
opti.solver('ipopt',{'ipopt.max_iter':100})

(h, g) = ca.hessian(opti.f, opti.x)
j = ca.jacobian(opti.g, opti.x)
plt.subplot(1, 2, 2)
opti.callback(lambda i: plt.semilogy(i,ca.np.linalg.cond(opti.debug.value(ca.vertcat(g.T, j)).todense()),'o'))


# Solve
sol = opti.solve_limited()
plt.title('U,X- V_guess ='+str(vg))
plt.grid(True)
plt.show()
```

**Comment by b-mare** [comment]
Hello,

Despite everything that is said above being correct suggestions for better solutions, that’s not really the issue in this example.

The main issue is that you did not define time to be positive.
If you just add
`opti.subject_to(tf>=0)`
both of your versions will converge just fine in a few iterations...

Here a more concise version
```
import casadi as ca

# -------------- System dynamics --------------
# Control
u = ca.MX.sym("u")

# State
x = ca.MX.sym("x", 3)
s = x[0]  # position
v = x[1]  # speed
m = x[2]  # mass

# ODE right hand side
sdot = v
vdot = (u - 0.2 * v**2) / m
mdot = -0.01 * u**2
xdot = ca.vertcat(sdot, vdot, mdot)

# ODE right hand side function
f = ca.Function("f", [x, u], [xdot])

# -------------- OCP --------------
N = 100
opti = ca.Opti()

# decision variables
X = opti.variable(3, N + 1)
s = X[0, :]
v = X[1, :]
m = X[2, :]

U = opti.variable(1, N)
T = opti.variable()

# objective
opti.minimize(T)  # Minimal time

# dynamic constraints
dt = T / N  # length of a control interval
for k in range(N):  # loop over control intervals
    # Explicit Euler integration over interval dt
    x_next = X[:, k] + dt * f(X[:, k], U[:, k])
    opti.subject_to(X[:, k + 1] == x_next)  # gap closing constraints

# path constraints
opti.subject_to(opti.bounded(-1.1, U, 1.1))
opti.subject_to(opti.bounded(0, v, 1.7))

# boundary conditions
opti.subject_to(s[0] == 0)  # start at position 0
opti.subject_to(v[0] == 0)  # with velocity zero
opti.subject_to(m[0] == 1)  # Initial mass

opti.subject_to(s[-1] == 10)  # Finish at position 10
opti.subject_to(v[-1] == 0)  # with velocity zero

opti.subject_to(T >= 0)  # Time must be positive

# initial values for solver
opti.set_initial(s, 1)
opti.set_initial(T, 1)
opti.set_initial(m, 1)

# solve NLP
opti.solver("ipopt")  # set numerical backend
sol = opti.solve()  # actual solve
```

And to answer on the Opti abstraction; yes you could define your variables inside of the loop, but it not really necessary in this example.

**Reply by sandeep026** [reply]
Wow. Such an important constraint got overlooked. Indeed it converges in a few iterations as @b-mare  suggested.

![image](https://github.com/casadi/casadi/assets/16457676/436dcc40-4d59-46f1-b319-d0c4e1116bab)

**Comment by jaeandersson** [comment]
Sounds like the issue has been resolved. Reopen if needed
